headlamp-plugin: Add package command

joaquimrocha opened the pull request:
This command creates a tarball in the format that headlamp-plugin expects to be in when we use the install command.

illume commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -1048,6 +1115,27 @@ yargs(process.argv.slice(2))
     }
   )
+  .command(
+    'pack <pluginPath> <outputDir>',
```
Can you please add docs to these spots?
- headlamp-plugin/README.md
- headlamp/docs/development/plugins/building.md

joaquimrocha replied:
I will add to README.md but for the building, we need more docs, as pack is tied to install as well.

illume commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -231,6 +233,71 @@ function extract(pluginPackagesPath, outputPlugins) {
+    return 1;
+  }
+
+  if (!fs.existsSync(outputFolderPath)) {
```
Maybe some of this code could be shared with code in extract(). This could be done later.

illume reviewed (changes requested):
Nice one. Left a few comments for your consideration.

illume commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -1048,6 +1115,27 @@ yargs(process.argv.slice(2))
   )
+  .command(
+    'pack <pluginPath> <outputDir>',
+    'Creates a tarball of the plugin package in the format Headlamp expects.',
```
I wonder if it should error before overwriting? I guess it would be similar to `npm pack`, and overwrite.

joaquimrocha replied:
I will make it err if the destination tarball exists.

illume commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -231,6 +233,71 @@ function extract(pluginPackagesPath, outputPlugins) {
   return 0;
 }
 
+/**
```
headlamp/plugins/headlamp-plugin/template/package.json is where you'd add it so it's added to peoples packages.

When you do `npm run pack` should it output in the current directory? Like `npm pack` does?

joaquimrocha replied:
yeah, I was wondering about what the package.json script should be named. I think 2 packs will be confusing?

yolossn commented:
Not a blocker but It would be nice to have the command also print the checksum that can be used in the `artifacthub-pkg.yaml`.

joaquimrocha commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -1048,6 +1115,27 @@ yargs(process.argv.slice(2))
+    'pack <pluginPath> <outputDir>',
+    'Creates a tarball of the plugin package in the format Headlamp expects.',
+    yargs => {
+      yargs.positional('pluginPath', {
```
Let's add it later.

joaquimrocha commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -1048,6 +1115,27 @@ yargs(process.argv.slice(2))
       process.exitCode = extract(argv.pluginPackages, argv.outputPlugins);
     }
   )
+  .command(
```
Done.

joaquimrocha commented:
Updated the branch with the checksum printing, test, and error when the final tarball already exists.

joaquimrocha commented:
I have updated it again:
- Now the command is called `package` instead of `pack` so we keep the latter for NPM
- Now the command defaults to the current dir in both options, like build does
- Added the package script to the template package.json

yolossn reviewed (commented):
Getting this error message but the tar.gz file is created

<img width="957" alt="image" src="https://github.com/user-attachments/assets/e26ca692-50d7-4240-a4ae-7de6f69772a5">

when i try to extract the tar.gz file I am getting this error
<img width="912" alt="image" src="https://github.com/user-attachments/assets/8e334381-dd5c-49fc-921c-db35fa49bfc3">

joaquimrocha commented:
@yolossn there was an error in extract as it didn't return the error code when no packages were extracted. I have updated that.

yolossn reviewed (approved):
Tested.

![image](https://github.com/user-attachments/assets/e3933fa6-02c6-47f4-ba83-8d9ec1afe498)

illume commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -226,11 +236,107 @@ function extract(pluginPackagesPath, outputPlugins) {
+ * Calculate the checksum of a file.
+ *
+ * @param {*} filePath
+ * @returns
```
The docs here are missing.

illume reviewed (approved):
👍 great!
